bridge/commands/doublepuppet.go

```diff
@@ -69,15 +79,53 @@ var CommandLogoutMatrix = &FullHandler{
 		Section:     HelpSectionAuth,
 		Description: "Disable double puppeting.",
 	},
-	RequiresLogin: true,
+	RequiresLogin:                 true,
+	RequiresMatrixLogin:           true,
+	RequiresManualDoublePuppeting: true,
+	RequiresNoAutoDoublePuppeting: true,
 }
 
 func fnLogoutMatrix(ce *Event) {
-	puppet := ce.User.GetIDoublePuppet()
-	if puppet == nil || puppet.CustomIntent() == nil {
-		ce.Reply("You don't have double puppeting enabled.")
+	ce.User.GetIDoublePuppet().ClearCustomMXID()
+	ce.Reply("Successfully disabled double puppeting.")
+}
+
+var CommandRefreshMatrix = &FullHandler{
+	Func: fnRefreshMatrix,
+	Name: "refresh-matrix",
+	Help: HelpMeta{
+		Section:     HelpSectionAuth,
+		Description: "Refresh double puppeting with an access token managed by the bridge.",
+	},
+	RequiresLogin:                  true,
+	RequiresRefreshableMatrixLogin: true,
+}
+
+func fnRefreshMatrix(ce *Event) {
+	if !ce.Bridge.DoublePuppet.CanAutoDoublePuppet(ce.User.GetMXID()) {
+		ce.Reply("This bridge instance has disabled automatic double puppeting for your Matrix server.")
 		return
 	}
-	puppet.ClearCustomMXID()
-	ce.Reply("Successfully disabled double puppeting.")
+	var err error
+	puppet := ce.User.GetIDoublePuppet()
+	if puppet != nil {
+		intent := puppet.CustomIntent()
+		if intent != nil && intent.SetAppServiceUserID {
+			ce.Reply("There is no need to refresh your double puppet, as it is currently managed by the bridge.")
+			return
+		}
+		puppet, ok := puppet.(bridge.RefreshableDoublePuppet)
+		if !ok {
+			ce.Reply("The bridge does not support refreshing your double puppet.")
+			return
+		}
+		err = puppet.RefreshCustomMXID()
```

Comment:
This new function seems like a lot more effort than just `SwitchCustomMXID` like it does below

Also not sure if this command should be added at all. The as_token method is the preferred way and the other methods will probably be removed in the future.